src/raftstore/coprocessor/region_info_accessor.rs

```diff
@@ -0,0 +1,753 @@
+                collided_region_id, region_id
+            );
+            self.regions.remove(&collided_region_id);
+        }
+        true
+    }
+
+    fn create_region(&mut self, region: Region) {
+        let end_key = data_end_key(region.get_end_key());
+        let region_id = region.get_id();
+
+        if !self.check_exist(region_id, region.get_region_epoch().get_version(), &end_key) {
+            warn!(
+                "region_collection: creating region {} but it's end_key collides with another \
+                 newer region.",
+                region.get_id()
+            );
+            return;
+        }
+
+        // Create new region
+        self.region_ranges.insert(end_key, region.get_id());
+
+        // TODO: Should we set it follower?
+        assert!(
+            self.regions
+                .insert(
+                    region.get_id(),
+                    RegionInfo::new(region, StateRole::Follower)
+                )
+                .is_none(),
+            "region_collection: trying to create new region {} but it already exists.",
+            region_id
+        );
+    }
+
+    fn update_region(&mut self, region: Region) {
+        let mut end_key_updated = false;
+
+        {
+            let existing_region_info = self.regions.get_mut(&region.get_id()).unwrap();
+            // Ignore stale messages or messages with nothing updating
+            {
+                let epoch = region.get_region_epoch();
+                let existing_epoch = existing_region_info.region.get_region_epoch();
+                if epoch.get_version() <= existing_epoch.get_version()
+                    && epoch.get_conf_ver() <= existing_epoch.get_conf_ver()
+                {
+                    return;
+                }
+            }
+
+            let old_region = &mut existing_region_info.region;
+            assert_eq!(old_region.get_id(), region.get_id());
+
+            // If the end_key changed, the old item in `region_ranges` should be removed.
+            if old_region.get_end_key() != region.get_end_key() {
+                end_key_updated = true;
+                // The region's end_key has changed.
+                // Remove the old entry in `self.region_ranges`.
+                let old_end_key = data_end_key(old_region.get_end_key());
+
+                let old_id = self.region_ranges.remove(&old_end_key).unwrap();
+                assert_eq!(old_id, region.get_id());
+            }
+
+            // If the region already exists, update it and keep the original role.
+            *old_region = region.clone();
+        }
+
+        // The new region info has a different end_key, so we need to update it in the
+        // `region_ranges` map.
+        if end_key_updated {
+            // If the region already exists, then `region_ranges` need to be updated.
+            // However, if the new_end_key has already mapped to another region, we should only
```

Comment:
Any test to cover this logic?

Reply:
Of course it's covered. In `test_extreme_cases`